### DISCUSSION TITLE
wa-sqlite IDBBatchAtomicVFS consistency and journal mode

### DISCUSSION THREAD

**Discussion by rhashimoto** [trong-orm/trong-orm#35] [General]
This PRAGMA to journal to memory raised a mental red flag when I saw it:

https://github.com/trong-orm/trong-orm/blob/e1da525157d14b6125c22f249e5c16998bc55d77/packages/wa-sqlite-web-backend/src/waSqliteWebBackend.ts#L49

This is safe against database corruption with IDBBatchAtomicVFS as long as SQLite is actually using batch atomic writes. If SQLite decides not to use batch atomic writes then an interruption during a write transaction can produce unrecoverable corruption when there is no persistent journal file.

SQLite will not use batch atomic writes for a write transaction involving multiple databases. Using the default journal mode is highly recommended if this type of transaction can happen. When SQLite _can_ use batch atomic writes, it writes the journal to memory as long as it fits in the cache. So with the default journal mode you will still get the performance benefit of a memory journal for "normal" transactions as long as the journal is smaller than the configured cache size (though the use of the cache for journal blocks will reduce the space for other cached blocks, so that's something to be aware of). If the journal does not fit into the cache then it will be moved to the file system as a fallback.

I bring this up because your project appears to allow the use of arbitrary SQL. If users will never use `ATTACH` to create or open another database file (or if you don't require persistence) then all this can be disregarded, but I wanted to make sure you were aware of the possible consequences if that happened.

**Comment by quolpr** [comment]
@rhashimoto wow, thank you so much for the explanation! Right now I am only experimenting, but I think I am not going to support `ATTACH`. 

> If SQLite decides not to use batch atomic writes then an interruption during a write transaction can produce unrecoverable corruption when there is no persistent journal file.

That is good to know 🤔

**Comment by quolpr** [comment]
@rhashimoto also, I am curious. Do you know, is it enough that `wa-sqlite-web-backend` will be under GPL 3 license, but the other codebase will remain MIT?

**Reply by rhashimoto** [reply]
I'm not qualified to answer that question, and I actually don't think that even software copyright experts can provide a definitive answer right now. This would likely have to be decided by court rulings and as far as I know, no one has taken it that far. As far as my understanding of the intent of the GPL goes, this would not be acceptable. You might find a better informed opinion on [Open Source Stack Exchange](https://opensource.stackexchange.com/) (or at least a variety to choose from 😀).

**Reply by rhashimoto** [reply]
The FSF's opinions are probably relevant in the absence of any court ruling:

* [Can I apply the GPL when writing a plug-in for a nonfree program?](https://www.gnu.org/licenses/gpl-faq.en.html#GPLPluginsInNF)
* [When is a program and its plug-ins considered a single combined program?](https://www.gnu.org/licenses/gpl-faq.en.html#GPLPlugins)

It looks like the FSF says that separately licensing a plug-in is at best borderline. A database that holds program state probably indicates a tighter relationship than that.
